### ISSUE TITLE
Attempt to provide a compiler plugin for writing Kotlin Builder boilerplate code.

### ISSUE THREAD

**Issue by Llonvne** [Llonvne/KBuilderDsl#1]
Kotlin's **data class** is a handy tool, but when a data class has too many properties, we often encounter some difficulties during initialization.You might have seen the following code.
```kotlin
val newCall = IrConstructorCallImpl(
    0,
    0,
    type = implClassSymbol.defaultType,
    symbol = newConstructor,
    0,
    0,
    0,
    null
)
```
If your IDE does not provide variable name hints, it can be almost unreadable.Let's turn to a simpler example.
```kotlin
data class User(
    val username: String,
    val password: String = "123456",
    val nickname: String? = "llonvne",
    val name: String = username,
    val enable: Boolean = false
)
```
This is a typical user data class.But when you need to handle it, putting all the processing flow into the primary constructor could make it unbearably bloated, so we often write the following code."
```kotlin
val userId = 1
val username = Database.UserDb.byId(userId)?.username ?: "UNKNOWN"
/**
 * fetch some data...
 */
val user = User(username, password, nickname, name, false)
```
This results in many one-time variables being defined in the code, and they are used only once in the constructor. We hope to simplify this process.Here is a possible way to simplify it.
```kotlin
val user = User.build {
    val userId = 1
    username = Database.UserDb.byId(userId)?.username ?: "Unknown"
    /**
     * ...
     */
}
```
I'm not sure if you find this approach aesthetically pleasing, I look forward to your feedback.
This way of writing has some obvious advantages: 
1. It saves unnecessary constructor calls. 
2. All object construction codes are contained within the build function body, making it semantically clear. 
3. It eliminates unnecessary variable definitions.
```kotlin
/**
 * The following code will be generated
 */
companion object {
    fun build(
        /**
         * The compiler will perform the following checks for userBuilderDsl,
         *
         * If a property is non-nullable in Dsl, no additional checks will be performed on the property
         * If a property is nullable in Dsl and the original property is nullable, no additional checks will be performed
         * If a property is nullable in Dsl but the original property is non-nullable, the following additional checks will be performed:
         * 1. The property must be assigned at least once
         * 2. Nullable values cannot be assigned to the property
         *
         * * If a variable is initialized based on previous properties,
         * and if the property remains null,
         * the check for this property will always be performed after the dependent property is checked.
         * If this property itself is nullable, then no check is performed; if this property is non-nullable and the dependent property is non-nullable, then no check is performed,
         * if this property is non-nullable but the dependent property is nullable, an exception is thrown
         * If a value has been assigned, then no check is performed
         */
        dsl: UserBuilderDsl.() -> Unit
    ): User {
        val builder = UserBuilderDsl()
        builder.dsl()
        return User(
            username = builder.username!!,
            password = builder.password,
            nickname = builder.nickname,
            name = builder.name!!,
            enable = builder.enable
        )
    }

    class UserBuilderDsl(
        /**
         * 1. All variables will be set to be reassignable, with nullability remaining consistent (unless explicitly modified below)
         * 2. If a variable is non-nullable and does not have an initial value, then the variable will be converted to nullable with an initial value of null in BuilderDSL
         * 3. If a variable has an initial value, then the variable will retain its original nullability and be given an initial value
         * 4. If a variable is initialized based on a previous property, it will become nullable with an initial value of null in BuilderDSL
         * Due to the above transformation rules, all parameters will have an initial value, meaning the DSL will have a no-argument constructor
         */
        // This variable is non-nullable without an initial value, it will be converted to nullable with an initial value of null
        var username: String? = null,
        // This variable is non-nullable with an initial value, it will retain its original nullability and have an initial value
        var password: String = "123456",
        var nickname: String? = "123456",

        // If a variable is initialized based on a previous property, then it will become nullable with an initial value of null
        var name: String? = null,
        var enable: Boolean = false
    )
}
/**
 * End of generated code
 */
```

**Comment by Llonvne** [comment]
中文的版本如下：
# 尝试提供编译器插件用于编写 Kotlin Builder 样板代码
Kotlin 的数据类(Data class) 是一个很好用的工具，但是当一个数据类具有过多的属性时，我们往往会在初始化时遇到一些困难，你可能见过如下的代码
```kotlin
val newCall = IrConstructorCallImpl(
    0,
    0,
    type = implClassSymbol.defaultType,
    symbol = newConstructor,
    0,
    0,
    0,
    null
)
```
如果你的IDE,没有提供变量名提示的话，他几乎难以阅读.让我们转向一个更简单的例子
```kotlin
data class User(
    val username: String,
    val password: String = "123456",
    val nickname: String? = "llonvne",
    val name: String = username,
    val enable: Boolean = false
)
```
这是一个典型的用户数据类,但是当你需要处理的时候，如果把所有的处理流程放到主构造器，可能导致主构造器臃肿不堪，所以我们经常会写出如下的代码
```kotlin
val userId = 1
val username = Database.UserDb.byId(userId)?.username ?: "UNKNOWN"
/**
 * fetch some data...
 */
val user = User(username, password, nickname, name, false)
```
这会让代码中定义许多一次性变量，而它们仅仅被使用一次在构造函数中，我们希望简化这个过程.下面是一种可能的简化方式
```kotlin
val user = User.build {
    val userId = 1
    username = Database.UserDb.byId(userId)?.username ?: "Unknown"
    /**
     * ...
     */
}
```
我不知道你们觉得这样的方式是否美观，我很期待来自你们的反馈.,这样的写法有一些显而易见的好处
1. 省去了多余的构造函数调用
2. 所有构造对象的代码均包含在 build 函数体内，语义清楚
3. 省去了不必要的变量定义

下面是生成代码的样子：
```kotlin
/**
 * 以下的代码将会被生成
 */
companion object {
    fun build(
        /**
         * 对于 userBuilderDsl 编译器将会作出如下检查，
         *
         * 如果在 Dsl 中该属性不可空，不对该属性作出额外检查
         * 如果在 Dsl 中该属性可空，且源属性可空，不作额外检查
         * 如果 Dsl 属性可空，源数据不可空作出以下额外检查
         * 1. 至少向该属性赋一次值
         * 2. 不可向该属性赋可空值
         *
         * * 如果一个变量依赖前面的属性进行初始化，
         * 如果该属性保持 null 值，
         * 该属性检查将始终在被依赖属性之后检查，
         * 如果该属性本身可空，那么不做任何检查，如果该属性非空，且依赖的属性也非空那么也不做检查，
         * 如果该属性非空，依赖的属性可空，抛出异常
         * 如果被赋值了则不做检查
         */
        dsl: UserBuilderDsl.() -> Unit
    ): User {
        val builder = UserBuilderDsl()
        builder.dsl()
        return User(
            username = builder.username!!,
            password = builder.password,
            nickname = builder.nickname,
            name = builder.name!!,
            enable = builder.enable
        )
    }

    class UserBuilderDsl(
        /**
         * 1. 所有变量都会被设置为可重新赋值，可空性保持一致（除非在下文有显式修改）
         * 2. 如果一个变量不具有初始值且不可空，那么该变量在 BuilderDSL 将会被转换为可空且具有初始值 null
         * 3. 如果一个变量具有初始值，那么该变量将会保持原本的可空性，并为其添加初始值
         * 4. 如果一个变量依赖前面的属性进行初始化，那么该变量在 BuilderDSL 将会变为可空，且具有 null 初始值
         * 由于上面规则的转化，所有的参数都有拥有一个初始值，即 DSL 将拥有无参数构造器
         */
        // 该变量不可空不具有初始值，将会被转换为可空且具有初始值 null
        var username: String? = null,
        // 该变量不可空具有初始值，将会保持原有可空性，且具有初始值
        var password: String = "123456",
        var nickname: String? = "123456",

        // 如果变量依赖前面的属性进行初始化，那么将变为可空且拥有 null 初始值
        var name: String? = null,
        var enable: Boolean = false
    )
}
/**
 * 生成的代码结束
 */
```

**Comment by Llonvne** [comment]
We can provide an annotation to instruct the compiler to generate this code. A possible annotation is as follows:
```kotlin
annotation class BuilderDsl

@BuilderDsl
data class User(
    val username: String,
    val password: String = "123456",
    val nickname: String? = "llonvne",
    val name: String = username,
    val enable: Boolean = false
) 
```
Since Lombok uses the Builder annotation, and our Builder has more of a DSL style, I would like to name it **BuilderDsl**.

**Comment by Llonvne** [comment]
Here are some points to consider: even though we try to provide some null safety checks, we still cannot ensure complete null safety.I have an idea to ensure the implementation of null safety. If a property is declared as non-nullable and does not have an initial value, we can provide a special function with the same name as the property that returns a non-nullable variable. At the same time, we make the nullable property in the DSL private.In this way, we can ensure null safety without relying on Kotlin's intelligent inference, but it inevitably introduces some complexity.

以下是一些值得思考的点，即便我们尝试提供一些空安全的检查，但是我们还是无法保证完整空安全，我有一个想法来保证空安全的实现，如果一个属性被声明为非空且不具备初始值，我们可以提供一个特殊的函数，函数名即为属性名，让他返回一个非空的变量即可，同时我们私有化DSL中原本那个可空的属性。这样我们可以不依赖Kotlin 智能推断，就可以保证空安全，但是也不可避免的带来一些复杂度

```kotlin
class User {
    companion object {
        fun build(
            dsl: UserBuilderDsl.() -> Unit
        ): User {
            val builder = UserBuilderDsl()
            builder.dsl()
            return User(
                username = builder.getUsername()!!
            )
        }

        class UserBuilderDsl(
            /**
             * 在这里我们私有化 username 变量
             */
            private var username: String? = null,
        ) {
            /**
             * 在这里我们提供一个 同名函数用于构建 username
             */
            fun username(
                build: () -> String
            ) {
                this.username = build()
            }

            /**
             * 提供一个 getter 用于取出 username
             */
            fun getUsername() = username
        }
    }
}
```

This almost leads to a similar problem as with the previous solution. We need to analyze the syntax to ensure that the function is called in all branches of the scope, which seems to be somewhat challenging.But this simplifies our issue of assigning nullable variables to source types that are non-nullable. We no longer need a custom compiler plugin to achieve this.

这几乎会带来与之前的解决方法类似的问题，我们需要分析语法来确保该函数在作用域所有分支都被调用，这似乎有点困难。但是这简化了我们对于像源类型不可空的变量赋值可空变量的问题，我们不再需要自定义编译器插件来实现

**Comment by Llonvne** [comment]
#### English
It seems the following is the code I hope to generate. Another point worth discussing is whether we should provide similar builders for all properties, or only for those non-nullable properties without initial values to indicate their uniqueness.
```kotlin
/**
 * The following code will be generated
 */
companion object {
    fun build(
        /**
         * For userBuilderDsl, the compiler will perform the following checks,
         * For properties that are non-nullable and have no initial value, ensure the build function is called at least once in all branches.
         * For properties that are initialized based on other properties, the above check is not applied.
         */
        dsl: UserBuilderDsl.() -> Unit
    ): User {
        val builder = UserBuilderDsl()
        builder.dsl()
        return User(
            username = builder.getUsername()!!,
            password = builder.password,
            nickname = builder.nickname,
            name = builder.getName()!!,
            enable = builder.enable
        )
    }

    class UserBuilderDsl(
        /**
         * 1. All variables will be set to be reassignable, and their nullability remains consistent (unless explicitly modified below).
         * 2. If a variable doesn’t have an initial value and is non-nullable, then this variable in BuilderDSL will be converted to nullable with an initial value of null.
         * 3. If a variable has an initial value, it will retain its original nullability and have an initial value added.
         * 4. If a variable is initialized based on a previous property, then this variable in BuilderDSL will become nullable with an initial value of null.
         * Due to the above rules of conversion, all parameters will have an initial value, meaning the DSL will have a no-argument constructor.
         */
        // This variable is non-nullable without an initial value, will be converted to nullable with an initial value of null.
        private var username: String? = null,
        // This variable is non-nullable with an initial value, will retain its original nullability and have an initial value.
        var password: String = "123456",
        var nickname: String? = "123456",

        // If the variable is initialized based on a previous property, it will become nullable with an initial value of null.
        private var name: String? = null,
        var enable: Boolean = false
    ) {
        fun username(
            build: () -> String
        ) {
            this.username = build()
        }

        fun getUsername() = username

        fun name(build: () -> String) {
            this.name = build()
        }

        fun getName() = name
    }
}
/**
 * End of generated code
 */
```
#### Chinese

似乎下面是我希望的能够生成的代码，还有一个值得讨论的点，是否应该为所有的属性提供类似的构建器呢，还是只是为非空的且没有初始值的属性提供来表示其特殊性
```kotlin
/**
 * 以下的代码将会被生成
 */
companion object {
    fun build(
        /**
         * 对于 userBuilderDsl 编译器将会作出如下检查，
         * 对于不可空且没有初始值的属性，在所有分支至少调用一次构建函数
         * 对于依赖于其他属性初始化的属性，没有上述检查
         */
        dsl: UserBuilderDsl.() -> Unit
    ): User {
        val builder = UserBuilderDsl()
        builder.dsl()
        return User(
            username = builder.getUsername()!!,
            password = builder.password,
            nickname = builder.nickname,
            name = builder.getName()!!,
            enable = builder.enable
        )
    }

    class UserBuilderDsl(
        /**
         * 1. 所有变量都会被设置为可重新赋值，可空性保持一致（除非在下文有显式修改）
         * 2. 如果一个变量不具有初始值且不可空，那么该变量在 BuilderDSL 将会被转换为可空且具有初始值 null
         * 3. 如果一个变量具有初始值，那么该变量将会保持原本的可空性，并为其添加初始值
         * 4. 如果一个变量依赖前面的属性进行初始化，那么该变量在 BuilderDSL 将会变为可空，且具有 null 初始值
         * 由于上面规则的转化，所有的参数都有拥有一个初始值，即 DSL 将拥有无参数构造器
         */
        // 该变量不可空不具有初始值，将会转换为可空且具有初始值 null
        private var username: String? = null,
        // 该变量不可空具有初始值，将会保持原有可空性，且具有初始值
        var password: String = "123456",
        var nickname: String? = "123456",

        // 如果变量依赖前面的属性进行初始化，那么将变为可空且拥有 null 初始值
        private var name: String? = null,
        var enable: Boolean = false
    ) {
        fun username(
            build: () -> String
        ) {
            this.username = build()
        }

        fun getUsername() = username

        fun name(build: () -> String) {
            this.name = build()
        }

        fun getName() = name
    }
}
/**
 * 生成的代码结束
 */
```

**Comment by Llonvne** [comment]
Another interesting syntax is that we can provide a top-level function to directly use User { /**building...*/ }. To prevent pollution of the top-level scope, we will provide an option to instruct the compiler whether to generate it or not.

另一个有趣的语法是我们可以提供顶层函数来直接使用 User{ /**building...*/ }，出于防止污染顶级作用域的考虑，我们将提供一个选项来指示编译器是否生成它
#### Definition
```kotlin
annotation class BuilderDsl(
    /**
     * This will instruct the compiler whether to generate a function with the same name as the type.
     */
    val topLevel: Boolean = false
)

/**
 * Constructor form of Builder
 * If topLevel is set to true,
 * the following constructor will be generated.
 */
fun User(dsl: User.Companion.UserBuilderDsl.() -> Unit): User {
    return User.build(dsl)
}
```
#### Usage
```kotlin
// Before
val user = User.build {
    val userId = 1
    username { "123" }
}

// After
val user2 = User {
    username { "123" }
}
```

**Comment by Llonvne** [comment]
Another issue is that, in many cases, the Builder can be passed around between scopes as an object, which poses a more serious challenge to the already complex null safety analysis. I am considering how to address this issue.
还有一个问题在于，在许多场合下 Builder 可以作为一个对象在作用域之间传递，这会对本就不容易的空安全分析带来更加严峻的挑战，我正在思考如何实现这个问题
```kotlin
val a = StringBuilder()
a.append("...")
/**
 * pass to a function or ...
 */
```

**Comment by Llonvne** [comment]
In addition, if we restrict the order of initialization, requiring users to initialize all non-nullable variables without default values at the beginning, this issue will become relatively simpler.
Consider the following builder pattern.
#### Definition
```kotlin
data class User(
    val username: String,
    val age: Int,
    val password: String = "123456",
    val nickname: String? = "llonvne",
    val name: String = username,
    val enable: Boolean = false
)
```
#### Usage
```kotlin
val user = User.builder()
    .username { "123" }
    .age { 12 }
    .build {
        nickname { "HelloWorld" }
    }
```
Let me start by explaining why such a design is needed. Under normal circumstances, the compiler will identify all non-nullable properties without default values inside a Data Class and generate the following interface.
```kotlin
fun interface UserUsernameBuilder {
    fun username(builder: () -> String): UserAgeBuilder
}

fun interface UserAgeBuilder {
    fun age(builder: () -> Int): UserBuilder
}
```
These interfaces only provide functions to initialize the current property and then point to the next interface that needs to be initialized. Only when there are no non-nullable properties without default values will it point to the general Builder.With this approach, we can fully ensure that all non-nullable properties without initial values are correctly initialized.
The implementation is relatively straightforward. All interfaces are implemented in the general builder, and the preceding interfaces only point to a single object, just temporarily hiding some methods.
```kotlin
class UserBuilder(
    private var username: String? = null,
    private var age: Int? = null,
    private var password: String = "123456",
    private var nickname: String? = "123456",
    private var name: String? = null,
    private var enable: Boolean = false
) : UserUsernameBuilder, UserAgeBuilder {
    override fun username(builder: () -> String): UserBuilder {
        this.username = builder()
        return this
    }

    override fun age(builder: () -> Int): UserBuilder {
        this.age = builder()
        return this
    }

    fun password(builder: () -> String): UserBuilder {
        this.password = builder()
        return this
    }

    fun nickname(builder: () -> String?): UserBuilder {
        this.nickname = builder()
        return this
    }

    fun name(builder: () -> String): UserBuilder {
        this.name = builder()
        return this
    }

    fun enable(builder: () -> Boolean): UserBuilder {
        this.enable = builder()
        return this
    }

    fun build(dsl: UserBuilder.() -> Unit): User {
        return User(
            username = this.username!!,
            age = this.age!!,
            password = this.password,
            nickname = this.nickname,
            name = this.name ?: username!!,
            enable = this.enable
        )
    }
}
```
This implementation brings some benefits: 
1. We always need only one Builder object, avoiding the overhead of intermediate objects. 
2. If we don't forcibly convert variables, the builder pattern will ensure that non-nullable properties without default values are properly initialized (if force conversion is used, it should be nullable and handled accordingly).

Of course, it also brings some problems. 
1. we need to generate a type for every non-nullable property without a default value, which could potentially lead to an increase in package size. 
2.  initialization must be carried out in the order in which non-nullable properties without default values are defined, which introduces some constraints.

I refer to this as the strict builder pattern.

```kotlin
annotation class StrictBuilderMode(
    val topLevel: Boolean = false
)

fun User(): UserUsernameBuilder {
    return User.builder()
}

/* user scope */
    companion object {
        fun builder(): UserUsernameBuilder {
            return UserBuilder()
        }
    }
```
The complete code will be demonstrated below.
```kotlin
fun main() {
    val user = User.builder()
        .username { "123" }
        .age { 12 }
        .build {
            nickname { "HelloWorld" }
        }
}

@StrictBuilderMode
data class User(
    val username: String,
    val age: Int,
    val password: String = "123456",
    val nickname: String? = "llonvne",
    val name: String = username,
    val enable: Boolean = false
)
/**
 * The following code will be generated
 */
{
    companion object {
        fun builder(): UserUsernameBuilder {
            return UserBuilder()
        }
    }
}

annotation class StrictBuilderMode(
    val topLevel: Boolean = false
)

fun User(): UserUsernameBuilder {
    return User.builder()
}


fun interface UserUsernameBuilder {
    fun username(builder: () -> String): UserAgeBuilder
}

fun interface UserAgeBuilder {
    fun age(builder: () -> Int): UserBuilder
}


class UserBuilder(
    private var username: String? = null,
    private var age: Int? = null,
    private var password: String = "123456",
    private var nickname: String? = "123456",
    private var name: String? = null,
    private var enable: Boolean = false
) : UserUsernameBuilder, UserAgeBuilder {
    override fun username(builder: () -> String): UserBuilder {
        this.username = builder()
        return this
    }

    override fun age(builder: () -> Int): UserBuilder {
        this.age = builder()
        return this
    }

    fun password(builder: () -> String): UserBuilder {
        this.password = builder()
        return this
    }

    fun nickname(builder: () -> String?): UserBuilder {
        this.nickname = builder()
        return this
    }

    fun name(builder: () -> String): UserBuilder {
        this.name = builder()
        return this
    }

    fun enable(builder: () -> Boolean): UserBuilder {
        this.enable = builder()
        return this
    }

    fun build(dsl: UserBuilder.() -> Unit): User {
        return User(
            username = this.username!!,
            age = this.age!!,
            password = this.password,
            nickname = this.name ?: username!!,
            name = this.name!!,
            enable = this.enable
        )
    }
}
```

**Comment by Llonvne** [comment]
### Concludion
I will prioritize implementing the builder pattern in strict mode, as it seems relatively simpler (I don’t have much information on how to implement the checks for calls in all branches). I will use KSP (Kotlin Symbol Processing) and Kotlin Poet to generate a bulk of code and will employ KCP to modify the internal definition of the User class."
